File: packages/api/src/handlers/rewards.ts
import { GetCommand, PutCommand, QueryCommand, UpdateCommand } from '@aws-sdk/lib-dynamodb';
import type { APIGatewayProxyResultV2 } from 'aws-lambda';
import { docClient, getTableName } from '../db';

const tableName = () => getTableName();

export async function listRewards(tenantId: string, programId: string): Promise<APIGatewayProxyResultV2> {
  const pk = `TENANT#${tenantId}#PROGRAM#${programId}`;
  const result = await docClient.send(
    new QueryCommand({
      TableName: tableName(),
      KeyConditionExpression: 'pk = :pk AND begins_with(sk, :sk)',
      ExpressionAttributeValues: { ':pk': pk, ':sk': 'REWARD#' },
    })
  );
  const items = (result.Items || []).map((item) => ({
    rewardId: item.sk?.replace('REWARD#', ''),
    name: item.name,
    pointsCost: item.pointsCost,
    tierEligibility: item.tierEligibility,
  }));
  return { statusCode: 200, headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ rewards: items }) };
}

export async function createReward(
  tenantId: string,
  programId: string,
  body: { name: string; pointsCost: number }
): Promise<APIGatewayProxyResultV2> {
  const rewardId = `rew_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`;
  const now = new Date().toISOString();
  const pk = `TENANT#${tenantId}#PROGRAM#${programId}`;
  const sk = `REWARD#${rewardId}`;
  await docClient.send(
    new PutCommand({
      TableName: tableName(),
      Item: {
        pk,
        sk,
        name: body.name ?? 'Reward',
        pointsCost: body.pointsCost ?? 0,
        tierEligibility: null,
        createdAt: now,
        updatedAt: now,
      },
    })
  );
  return {
    statusCode: 201,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ rewardId, name: body.name ?? 'Reward', pointsCost: body.pointsCost ?? 0 }),
  };
}

export async function redeem(
  tenantId: string,
  programId: string,
  body: { memberId: string; rewardId: string }
): Promise<APIGatewayProxyResultV2> {
  const { memberId, rewardId } = body;
  if (!memberId || !rewardId)
    return { statusCode: 400, body: JSON.stringify({ error: 'memberId and rewardId required' }) };
  const pk = `TENANT#${tenantId}#PROGRAM#${programId}`;
  const rewardRes = await docClient.send(new GetCommand({ TableName: tableName(), Key: { pk, sk: `REWARD#${rewardId}` } }));
  const reward = rewardRes.Item;
  if (!reward) return { statusCode: 404, body: JSON.stringify({ error: 'Reward not found' }) };
  const cost = reward.pointsCost ?? 0;
  const balanceSk = `BALANCE#${memberId}`;
  const balanceRes = await docClient.send(new GetCommand({ TableName: tableName(), Key: { pk, sk: balanceSk } }));
  const current = balanceRes.Item?.points ?? 0;
  if (current < cost)
    return { statusCode: 400, body: JSON.stringify({ error: 'Insufficient points', balance: current, required: cost }) };
  const now = new Date().toISOString();
  const txnId = `txn_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`;
  const skTxn = `TXN#${now}#${txnId}`;

  await docClient.send(
    new UpdateCommand({
      TableName: tableName(),
      Key: { pk, sk: balanceSk },
      UpdateExpression: 'SET points = points - :cost, updatedAt = :now',
      ExpressionAttributeValues: { ':cost': cost, ':now': now },
    })
  );
  await docClient.send(
    new PutCommand({
      TableName: tableName(),
      Item: {
        pk,
        sk: skTxn,
        type: 'redemption',
        memberId,
        rewardId,
        points: cost,
        createdAt: now,
        gsi1pk: `TENANT#${tenantId}`,
        gsi1sk: `TXN#${programId}#${now}#${txnId}`,
      },
    })
  );
  return {
    statusCode: 200,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ transactionId: txnId, rewardId, pointsDeducted: cost, balance: current - cost }),
  };
}

File: packages/backend/src/app/services/rewards.py
"""Rewards list, create, redeem."""

import time
import random
import string
from botocore.exceptions import ClientError
from app.db import get_table, get_doc_client, get_table_name, serialize_item, key
from app.exceptions import BadRequestError, NotFoundError

_TXN_TTL_SECONDS = 18 * 30 * 24 * 3600


def _table():
    return get_table()


def _txn_ttl() -> int:
    return int(time.time()) + _TXN_TTL_SECONDS


def _reward_id() -> str:
    return f"rew_{int(time.time()*1000)}_{''.join(random.choices(string.ascii_lowercase + string.digits, k=9))}"


def _txn_id() -> str:
    return f"txn_{int(time.time()*1000)}_{''.join(random.choices(string.ascii_lowercase + string.digits, k=9))}"


def list_rewards(tenant_id: str, program_id: str) -> list[dict]:
    pk = key.program_scoped_pk(tenant_id, program_id)
    result = _table().query(
        KeyConditionExpression="pk = :pk AND begins_with(sk, :sk)",
        ExpressionAttributeValues={":pk": pk, ":sk": key.REWARD_SK_PREFIX},
    )
    items = result.get("Items", [])
    return [
        {
            "rewardId": it["sk"].replace(key.REWARD_SK_PREFIX, ""),
            "name": it.get("name", "Reward"),
            "pointsCost": it.get("pointsCost", 0),
            "tierEligibility": it.get("tierEligibility"),
        }
        for it in items
    ]


def create_reward(tenant_id: str, program_id: str, name: str, points_cost: int) -> dict:
    reward_id = _reward_id()
    now = time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime())
    pk = key.program_scoped_pk(tenant_id, program_id)
    sk = key.reward_sk(reward_id)
    _table().put_item(
        Item={
            "pk": pk,
            "sk": sk,
            "name": name,
            "pointsCost": points_cost,
            "tierEligibility": None,
            "createdAt": now,
            "updatedAt": now,
        },
    )
    return {"rewardId": reward_id, "name": name, "pointsCost": points_cost}


def redeem(tenant_id: str, program_id: str, member_id: str, reward_id: str) -> dict:
    pk = key.program_scoped_pk(tenant_id, program_id)
    tbl = _table()
    reward_res = tbl.get_item(Key={"pk": pk, "sk": key.reward_sk(reward_id)})
    reward = reward_res.get("Item")
    if not reward:
        raise NotFoundError("Reward not found")
    cost = int(reward.get("pointsCost", 0))
    sk_balance = key.balance_sk(member_id)

    # Early check for better error message (not the real guard)
    balance_res = tbl.get_item(Key={"pk": pk, "sk": sk_balance})
    current = int(balance_res.get("Item", {}).get("points", 0))
    if current < cost:
        raise BadRequestError(f"Insufficient points (balance={current}, required={cost})")

    now = time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime())
    txn_id = _txn_id()
    sk_txn = key.txn_sk(now, txn_id)
    table_name = get_table_name()

    try:
        get_doc_client().transact_write_items(TransactItems=[
            {
                "Update": {
                    "TableName": table_name,
                    "Key": {"pk": {"S": pk}, "sk": {"S": sk_balance}},
                    "UpdateExpression": "SET points = points - :cost, updatedAt = :now",
                    "ConditionExpression": "points >= :cost",
                    "ExpressionAttributeValues": {
                        ":cost": {"N": str(cost)},
                        ":now": {"S": now},
                    },
                }
            },
            {
                "Put": {
                    "TableName": table_name,
                    "Item": serialize_item({
                        "pk": pk,
                        "sk": sk_txn,
                        "type": "redemption",
                        "memberId": member_id,
                        "points": cost,
                        "rewardId": reward_id,
                        "createdAt": now,
                        "ttl": _txn_ttl(),
                        "gsi1pk": key.gsi1_tenant(tenant_id),
                        "gsi1sk": key.gsi1_txn_sk(program_id, now, txn_id),
                    }),
                }
            },
        ])
    except ClientError as e:
        if e.response["Error"]["Code"] == "TransactionCanceledException":
            raise BadRequestError(f"Insufficient points (balance={current}, required={cost})")
        raise

    return {"transactionId": txn_id, "rewardId": reward_id, "pointsDeducted": cost, "balance": current - cost}
